### DISCUSSION TITLE
nfcapd capture 10Million per second?

### DISCUSSION THREAD

**Discussion by jhavinit** [phaag/nfdump#562] [Q&A]
Hi folks,
We are planning to export 10Million per second netflow from router. Can a single instance of nfcapd can capture 10million per seconds netflows data if so what configs we can do. Can we do some sort of load balancing to split load to virtual interfaces and create a listener using each nfcapd, if so how? So far I have tried in ubuntu 64 core machine with 300gb ram to achieve ~1.5million per second. Please advise. Thanks in advance.

**Comment by phaag** [comment]
Hey @jhavinit,
THat's not an easy question. Do you expect the 10M flows from a single exporter or multiple exporters? Are the 10M flows the peek rate and a temporary maximum or is it a sustained packet rate with 10M?

You would need to identify the bottleneck and there are a few options:
1. Try to increase the number of workers (`-W`) to compress and store away your flows. 
2. The flow decoding may be the bottleneck: If you have netflow v9 or IPFIX, that needs more CPU than v5.
3. If you have multiple exporters sending v9 or ipfix, try to split them of different nfcapd processes, listening on different ports. 
4. The kernel may be the bottleneck. That may be a trickier one to solve.

There is for sure still some capacity to optimise nfdump, but without a detailed analysis on the object itself, it's rather guessing.
Please let me know, if I could help, if you have some ideas.

**Reply by jhavinit** [reply]
Hi @phaag Thanks for answering...
10Million per second is peak traffic.. but it can be sustained for few hours.
Iam spliiting the traffic in 2 routers/exporters using 6million and 4million exports.
We have v9 packets. I will try to use more workers to test it and update. Do we have any upper cap or benchmarking results for nfcapd? Also any ideas on load balancing ? Shall we try at network level or at application level.. I will try using nginx/haproxy though

**Comment by phaag** [comment]
More Workers only help, if compressing and storing away the data is the bottleneck. If v9 decoding is the bottleneck, it does not help. Decoding v9 is single threaded. The performance depends on how many elements are collected per flow.
Use two different `nfcapd` processes for both of your exporters. 
I have no benchmarks for the collector.
nginx/haproxy is not involved in any collection, so I don't know, what you mean here.

**Reply by jhavinit** [reply]
Ok got it. I meant i will use nginx to split the data collected from exporter on the port to multiple internal ports and have nfcapd listener for each port. Thanks
